### PR TITLE
Timer object

### PR DESCRIPTION
**Pull request by lkotipal**
Replace all phiprof timer calls with timer objects as per https://github.com/fmihpc/phiprof/pull/30. Timers are closed by their destructor e.g. when going out of scope, making it easier to avoid bugs caused by dangling timers.

Some guidelines:
- Never use `phiprof::start(args)` or `phiprof::stop(label, args)`, use `phiprof::Timer t {args}` and `t.stop(args)` instead.
- Use `int id = phiprof::initialize(args)` outside a loop and `phiprof::Timer {id}` inside. Initialization is superfluous if done immediately before starting a timer.
- Since timers automatically stop when going out of scope, you may omit `t.stop()` before e.g. return.

### CODE REVIEW

**Comment by lkotipal** [comment]
aaa

**Comment by ursg** [comment]
This is nice (and apparently works well), but I would suggest that the timers don't get variable names like "mpi", "communication" etc.
Because in its current form, stopping these timers then adds code lines like ```mpi.stop()``` or ```communication.stop()``` that, by themselves, make it in no way apparent that they are dealing with timers.

So either rename all of them to ```mpiTimer```,```communicationTimer``` etc, or change the name of the stop method to ```stopTimer```.

**Comment by lkotipal** [comment]
Normally I'd say the type conveys that information sufficiently but IntelliSense is a bit finicky over remote connections. I'll go through and give the timers better names.

**Comment by lkotipal** [comment]
Yolo'd this in regex, let's see if it compiles

**Comment by lkotipal** [comment]
Automatic testpackage output for pull request #721.

Based on commit 6e9c9be62d160cbecf8dd64189a5748883512849.

Process started by user ```lkotipal``` on host ```turso01.cs.helsinki.fi``` at Mon, 11 Sep 2023 14:59:52 +0300.

:heavy_check_mark: Compilation succeeded with 23 warnings.

:heavy_check_mark: Testpackage ran successfully.

:heavy_check_mark: acctest_2_maxw_500k_100k_20kms_10deg: Zero diffs.
:heavy_check_mark: acctest_3_substeps: Zero diffs.
:heavy_check_mark: transtest_2_maxw_500k_100k_20kms_20x20: Zero diffs.
:heavy_check_mark: acctest_4_helium: Zero diffs.
:heavy_check_mark: acctest_5_proton_antiproton: Zero diffs.
:heavy_check_mark: restart_write: Zero diffs.
:heavy_check_mark: restart_read: Zero diffs.
:large_orange_diamond: Magnetosphere_small: Nonzero diffs: `proton/vg_v_0` has absolute error 32255.9, relative error 0.0639677.
:large_orange_diamond: Magnetosphere_polar_small: Nonzero diffs: `proton/vg_v_nonthermal_0` has absolute error 102622, relative error 0.128336.
:heavy_check_mark: test_fp_fsolver_only_3D: Zero diffs.
:heavy_check_mark: test_fp_substeps: Zero diffs.
:heavy_check_mark: Flowthrough_trans_periodic: Zero diffs.
:heavy_check_mark: Flowthrough_x_inflow_y_outflow: Zero diffs.
:heavy_check_mark: Flowthrough_x_inflow_y_outflow_acc: Zero diffs.
:large_orange_diamond: Flowthrough_amr: Nonzero diffs: `proton/vg_rho_0` has absolute error 0.00289569, relative error 1.34239e-09.
:large_orange_diamond: Magnetosphere_3D_small: Nonzero diffs: `proton/vg_v_0` has absolute error 262.342, relative error 0.000349676.
:heavy_check_mark: Ionosphere_small: Zero diffs.

Full testpackage run output is available in `/wrk-vakka/users/lkotipal/testpackage/testpackage_run_6e9c9be62d160cbecf8dd64189a5748883512849.log` on host `turso01.cs.helsinki.fi`

**Comment by lkotipal** [comment]
Turso is having a normal one so here's a manual test

**Comment by lkotipal** [comment]
You're right, my references are probably outdated. I'll generate new ones and re-run.

File: vlasovsolver/vlasovmover.cpp [hunk 1]
```diff
@@ -455,73 +437,70 @@ void calculateAcceleration(dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>&
       for (uint popID=0; popID<getObjectWrapper().particleSpecies.size(); ++popID) {
          adjustVelocityBlocks(mpiGrid, cells, true, popID);
       }
+   } else {
+      // Fairly ugly but no goto
```
Comment (ursg) [inline vlasovsolver/vlasovmover.cpp:441]: This part is the only actual change of code semantics, other than phiprof variable reordering.
The testpackage result shows that this is working all as intended, but was this change required for the timer objects, or was it just something that irked you and you wanted to go away?

**Review by ursg — APPROVED** [review]
Great stuff! Scrolled through the whole diff just to be sure.
I also approve the occasional formatting fix that you put in.

Your initializer syntax of
```
int someTime {phiprof::initializeTimer("asdf")};
```
made me do a double-take in some places, but that's all ok!

File: vlasovsolver/vlasovmover.cpp [hunk 2]
```diff
@@ -272,9 +256,11 @@ void calculateSpatialTranslation(
    
    // If dt=0 we are either initializing or distribution functions are not translated. 
    // In both cases go to the end of this function and calculate the moments.
-   if (dt == 0.0) goto momentCalculation;
+   if (dt == 0.0) {
+      calculateMoments_R(mpiGrid,localCells,true);
+   }
```
Comment (markusbattarbee) [inline vlasovsolver/vlasovmover.cpp:261]: I would add a return here so we don't do a pointless zero-dt translation computation. Then it'll return to the same behaviour we had before.

Comment (lkotipal) [reply vlasovsolver/vlasovmover.cpp:261]: That was the intention, I think I goofed up the merge there

File: vlasovsolver/vlasovmover.cpp (hunk 1 above)
Comment (lkotipal) [reply vlasovsolver/vlasovmover.cpp:442]: iirc that actually didn't compile with the goto since it was jumping over a variable declaration

Comment (ursg) [reply vlasovsolver/vlasovmover.cpp:442]: Fair enough, that's definitely a sufficient reason to restructure.
